crates/blockchain/payload.rs

```diff
@@ -277,6 +313,75 @@ impl From<PayloadBuildContext> for PayloadBuildResult {
 }
 
 impl Blockchain {
+    /// Attempts to fetch a payload given it's id. If the payload is still being built, it will be finished.
+    /// Fails if there is no payload or active payload build task for the given id.
+    pub async fn get_payload(&self, payload_id: u64) -> Result<PayloadBuildResult, ChainError> {
+        let mut payloads = self.payloads.lock().await;
+        // Find the given payload and finish the active build process if needed
+        let idx = payloads
+            .iter()
+            .position(|(id, _)| id == &payload_id)
+            .ok_or(ChainError::UnknownPayload)?;
+        let finished_payload = (payload_id, payloads.remove(idx).1.to_payload().await?);
+        payloads.insert(idx, finished_payload);
+        // Return the held payload
+        match &payloads[idx].1 {
+            PayloadOrTask::Payload(payload) => Ok(*payload.clone()),
+            // We already converted the payload into a finished version
+            _ => unreachable!(),
+        }
+    }
+
+    /// Starts a payload build process. The built payload can be retrieved by calling `get_payload`.
+    /// The build process will run for the full block building timeslot or until `get_payload` is called
+    pub async fn initiate_payload_build(self: Arc<Blockchain>, payload: Block, payload_id: u64) {
+        let self_clone = self.clone();
+        let cancel_token = CancellationToken::new();
+        let cancel_token_clone = cancel_token.clone();
+        let payload_build_task = tokio::task::spawn(async move {
+            self_clone
+                .build_payload_loop(payload, cancel_token_clone)
+                .await
+        });
+        let mut payloads = self.payloads.lock().await;
+        if payloads.len() >= MAX_PAYLOADS {
+            // Remove oldest unclaimed payload
+            payloads.remove(0);
+        }
```

Comment:
This could remove ongoing building tasks.

Reply:
Is that a problem?